### PR TITLE
Feat: use extension-native APIs for app <-> wallet messaging

### LINKED ISSUE #776: Mozilla firefox issue when trying to use connect to login to the explorer app
Mozilla firefox issue when trying to use connect to login to explorer, I can login to explorer with chrome without an issue but cannot login with firefox

Video of issue http://somup.com/crVIVHol7x

If I am logged in already I get this screen but the links are not clickable and I am stuck on the page

Screen Shot 2021-01-12 at 8 50 31 PM

If I login with my seed phrase I get this screen with an infinite spinner

Screen Shot 2021-01-12 at 8 50 08 PM

### PR DESCRIPTION
**Pull request by hstove**
> Try out this version of the Stacks Wallet - download [extension builds](https://github.com/blockstack/ux/actions/runs/522097463), the [hosted version](https://pr-930.app.stacks.engineering), or the [Stacks testnet demo app](https://pr-930.testnet-demo.stacks.engineering)

Previously, we had design constraints in order to make authentication work with a hosted version of our authenticator. For security reasons, we've moved to an extension-only approach. This allows us to refactor our code to use APIs that only an extension can use. This greatly reduces complexity and flakiness of any popup-based approach.

For example, improvements include:

- No "popup blocked" issues in Firefox
- Instant response to the app after auth / transaction complete (instead of polling)
- Fully typed messaging APIs

It also extends the `StacksProvider` interface injected by the extension. The new APIs include:

```ts
/**
   * Make a transaction request
   * 
   * @param payload - a JSON web token representing a transaction request
   */
  transactionRequest(payload: string): Promise<FinishedTxPayload>;
  /**
   * Make an authentication request
   * 
   * @param payload - a JSON web token representing an auth request
   *
   * @returns an authResponse string in the form of a JSON web token
   */
  authenticationRequest(payload: string): Promise<string>;
  getProductInfo: undefined | (() => {
    version: string;
    name: string;
  });
```

Which is light and flexible enough to be implemented by future wallets, such as a native app with an integrated web browser.

### CODE REVIEW

File: packages/connect/src/popup.ts [hunk 1]
```diff
@@ -96,7 +100,9 @@ export const setupListener = <T>({
   // Send a message to the authenticator popup at a consistent interval. This allows
   // the authenticator to 'respond'.
   const pingInterval = 250;
-  const interval = setInterval(() => {
+  let interval: number | undefined = undefined;
```
Comment (kyranjamie) [inline packages/connect/src/popup.ts:103]: Could tidy this up as a recursive promise

File: packages/app/src/extension/background.ts [hunk 2]
```diff
@@ -2,17 +2,52 @@ import { wrapStore } from 'webext-redux';
...
+        break;
+      default:
+        break;
+    }
+  });
+});
+
+if (NODE_ENV === 'test') {
+  // expose a helper function to open a new tab with the wallet from tests
+  (window as any).openOptionsPage = function (page: string) {
```
[the remark this answers is not in the document]
Comment (hstove) [reply packages/app/src/extension/background.ts:44]: You're right about the type! This function is only ever used via Playwright, which is why it probably seems weird. There is no built-in function to "open options back of extension" in Playwright. https://github.com/blockstack/ux/blob/feat/extension-messaging/packages/app/tests/integration/page-objects/wallet.page.ts#L62:L62

File: packages/app/src/extension/message-types.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+export type MessageFromContentScript = AuthenticationRequestMessage | TransactionRequestMessage;
+
+export type MessageToContentScript = AuthenticationResponseMessage | TransactionResponseMessage;
+
+/**
+ * Inpage script <-> Content script
+ */
+
+export enum DomEventName {
```
[the remark this answers is not in the document]
Comment (hstove) [reply packages/app/src/extension/message-types.ts:54]: They are real DOM events

File: packages/connect/src/popup.ts (hunk 1 above)
Comment (hstove) [reply packages/connect/src/popup.ts:103]: Ah, this code needs to go :+1:

**Pull request closed without merging**